C++: Add bcopy to models and use it.

geoffw0 opened the pull request:
Add `bcopy` to the `MemcpyFunction` model.

Have `callDereferences` use models rather than duplicating this data itself.  @MathiasVP do you have any idea what `Dereferences.qll`'s `callDereferences` is used for?  I'd like to figure out a way to explicitly test this change.

MathiasVP commented:
> Have `callDereferences` use models rather than duplicating this data itself. @MathiasVP do you have any idea what `Dereferences.qll`'s `callDereferences` is used for? I'd like to figure out a way to explicitly test this change.

Not exactly sure, no. A quick search only found it being imported in `semmle/code/cpp/controlflow/Dataflow.qll`, where its functionality seems to be used by `killedBy`.

rdmarsh2 commented:
Looks like the only use is in `Critical/MissingNullTest.ql`.

geoffw0 commented:
> Looks like the only use is in Critical/MissingNullTest.ql.

Ah good spot, I'd missed this use.  I've added a test of MissingNullTest.ql that includes `bcopy`.

geoffw0 commented:
It looks like there's a failing test.  I will have to come back to this and investigate.
```
semmlecode-cpp-tests/DO_NOT_DISTRIBUTE/security-tests/CWE-119/VulnC/tests/MissingNullTest.qlref

-| bad_read2.c:36:15:36:18 | buff | Value may be null; it should be checked before dereferencing. |
 | strlen_missing_null.c:32:9:32:16 | new_buff | Value may be null; it should be checked before dereferencing. |
```

geoffw0 commented:
Sorry for the long delay getting back to this...

I've fixed the failing test - `callDereferences` still needed a special case for the varargs parameters to `printf` variants.

I've also modernized some code and removed the taint flow from size parameters; we don't generally consider that taint any more.

geoffw0 commented:
Fixed merge conflict.

rdmarsh2 commented:
Looks like `localtaint.ql` is failing now

geoffw0 commented:
Fixed.

rdmarsh2 commented on cpp/ql/src/semmle/code/cpp/models/implementations/Memcpy.qll:
```diff
@@ -13,61 +13,73 @@ import semmle.code.cpp.models.interfaces.Taint
  * The standard functions `memcpy` and `memmove`, and the gcc variant
```
This comment needs an update to include `bcopy`.

geoffw0 replied:
Done.

MathiasVP reviewed (approved):
LGTM!
